### PR TITLE
Reorganize to allow for future use through ESM (TECH-2322)

### PR DESCRIPTION
**Pull request by dtognazzini**
# Objective
* The original aim of this work is to provide source ESM for consumers to use.
* In support of the original aim, the internals were refactored to use ESM and reorganized to steer consumers away from using the namespacing methods for access.
* Upon further consideration, our main consumer of Backdraft will need to function in a hybrid world whereby some JS can access Backdraft modules via ESM while others continue to access via the namespacing methods.
* Supporting ESM access and namespacing access using the same underlying code, while possible, requires a lot of bundler configuration (Webpack) and results in the test environment differing significantly from dev/production environments (i.e. envs providing Backdraft as an external).
* We'd prefer to have our tests run against the same configuration as used in production. This requires all the code to use the greatest common denominator: namespacing access.
* Our main consumer will use ESM shims to provide a level of indirection for in-repo consumers. Once converting a critical mass worth of consumers over to using ESMs, we can fully use a Backdraft library providing ESMs.

# Summary
* Internals converted to ESM.
* Tests added to verify ESM implementations.
* Consumers wanting ESM access can get at the underlying classes via the `Backdraft` object. For example, to access the base view class:
```javascript
const View = Backdraft.View;
```
* Classes have been updated to allow for ES6 class extension (discussed below). Prior to these changes, consumers had to access the internals via the namespacing methods because the namespacing methods called things like `finalize` that is necessary for the internal code to work.

# Changes

## Backbone vs ES6 class semantics
Many Backdraft/Backbone classes were converted to ES6 classes. ES6 class syntax requires `super` to be called before doing anything else. Some Backdraft classes had constructors that run some code before calling super. This code was moved around on a case by case basis (described below) in an effort to maintain the same behavior.

Generally, all classes were converted away from using Backbone's `initialize` method. Any subclasses that implement an `initialize` method that relies on properties from the constructor will need to be changed to move that dependent code into the `constructor` just after `super` is called. Generally, `initialize` methods should be abandoned in favor of `constructor`.

### More Detail

The base class used in Backbone calls `initialize`. Properties that were previously initialized in the constructor will not be available in `initialize`.

For example, consider this ES5 Backbone class:

```javascript
var MyView = Backbone.View.extend({
  constructor: function(options) {
    this.service = options.service;
    MyView.__super__.constructor.apply(this, arguments);
  },

  initialize: function() {
    this.service.doSomething();
  }
});
```

Converting to ES6, we have to set the `service` property after calling `super`:

```javascript
class MyView extends Backbone.View {
  constructor(options) {
    super(options);

    this.service = options.service;
  },

  initialize() {
    this.service.doSomething();
  }
}
```

This code will break because `super` invokes Backbone's base class functionality of calling `initialize` and `initialize` expects `this.service` to have been set.

## Deleted
`Backdraft.Utils.DomVisibility` - internal helper, not provided publicly.
`Backdraft.Utils.Coordinates` - internal helper, not provided publicly.
`Backdraft.Utils.log`
`Backdraft.plugin.destroyAll`
`Backdraft.Utils.Class`

### `Plugin`
No longer extends Utils.Class.

Plugins can no longer assume any other plugins, include the deprecated Base plugin, have been installed. Plugins should ensure necessary plugins are installed via calling `app.installPlugin("NameOfPlugin")` in an initializer that requires that a plugin is installed.

### `App`
No longer extends Utils.Class.

`App.instances` - deleted.
`App.factory()` - deleted.
`appInstance.plugins` - deleted.

The preferred way of registering apps is now:

```javascript
const app = new Backdraft.App();
Backdraft.app("myapp", app);
```

To specify plugins, either pass an array into the constructor:
```javascript
const app = new Backdraft.App(["DataTable", "WhizBang"]);
Backdraft.app("myapp", app);
```

...or subclass and provide a `plugins` property:

```javascript
// via class property
class MyApp extends Backdraft.App {
  get plugins() {
    return ["DataTable", "WhizBang"];
  }
}
Backdraft.app("myapp", new MyApp());

// via prototype
class MyApp2 extends Backdraft.App { }
MyApp2.prototype.plugins = ["DataTable", "WhizBang"];
Backdraft.app("myapp2", new MyApp2());
```

*Note: The Base plugin is not installed when registering an App via an App instance.*

To maintain backwards compatibility, the Base plugin is installed when registering app using a prototype:
```javascript
const app = Backdraft.app("name", {
  plugins: ["DataTable", "WhizBang"]
});
app.installedPlugins
// -> ["Base", "DataTable", "WhizBang"]
```

## Base plugin
The Base plugin is deprecated. See the Deprecations section for more information.

### `Cache`
No longer extends Utils.Class.

### `View`
The `constructor` initialized the `children` property before calling super. `children` is reimplemented as a getter with lazy initialization.

### `Router`
The `constructor` initialized the `nameHelper` and `$el` properties before calling super. `nameHelper` was reimplemented as a getter with lazy initialization. Initializing `$el` was moved to happen after `super`. This means that `this.$el` is not available in `initialize`. After speaking with Nick, we think this is ok.

### `Item.finalize`
Deleted. It's an empty function and there are no known callers.

### `List`
The `constructor` initialized properties `options`, `cache` and `itemClass` before calling `super`. The initialization of `options` and `cache` was moved to `constructor`. This means that subclasses cannot access `this.cache` or `this.options` in `initialize`.

`itemClass` was converted to be a requirement for subclassing. Subclasses must define an `itemClass` getter or prototype property that is set to a `View` class.

`List.finalize` was internalized, deleted publicly. `view.listing.list()` requires that an `itemClassName` property is passed. The internalized version of `List.finalize` meets the new requirement of an `itemClass` getter by creating one on the fly from the `itemClassName` provided to `view.listing.list()`.

## DataTable Plugin
The internal classes used by the DataTable plugin are now exported to consumers. However, changes were made to these classes that change the contract for subclasses.

### `Row`
Newly exported class. `Row` is the base class used by `view.dataTable.row()` internally.

`Row.finalize` has been deleted.

### `LocalDataTable`
Newly exported class. `LocalDataTable` is the base class used by `view.dataTable()` internally.

The `LocalDataTable.finalize` has been deleted.

#### Changes Affecting Namespacing and ESM access
The `constructor` was initializing all of the following before calling `super`:

```javascript
_.extend(this, _.pick(this.options, [ "selectedIds" ]));
_.bindAll(this, "_onRowCreated", "_onBulkHeaderClick", "_onBulkRowClick", "_bulkCheckboxAdjust", "_onDraw",
    "_onColumnVisibilityChange", "_onColumnReorder");
this.cache = new Base.Cache();
this.selectionManager = new SelectionManager();
this.rowClass = this.options.rowClass || this._resolveRowClass();
this._applyDefaults();
this._columnManager = new ColumnManager(this);
this._lockManager = new LockManager(this);
```

All of this code was moved to be after the `super` call.

The `_triggerGlobalEvent` method has been deleted.

#### Changes Affecting ESM access
The `rowClass` parameter must be supplied to the constructor.

To specify column types, provide a `Backdraft.DataTable.Config` instance with the appropriate `Backdraft.DataTable.ColumnType`s to the `LocalDataTable`, for example:

```javascript
function createDatatableConfig() {
  const config = new Backdraft.DataTable.Config();
  config.addColumnType(new Backdraft.DataTable.ColumnType());
  return config;
}

class MyDataTable extends Backdraft.DataTable.Local {
  constructor(options) {
    super(_.extend({
      config: createDatatableConfig()
    }, options) );
  }
}
```

### `ServerSideDataTable`
`ServerSideDataTable` is the base class used by `view.dataTable()` when passing `serverSide: true`. All the changes affecting `LocalDataTable` also affect `ServerDataTable`.

The `ServerSideDataTable.finalize` inherited from `LocalDataTable.finalize` is no longer available.

#### Changes Affecting Namespacing and ESM access

The `constructor` initialized the `paginate` property before calling super. The initialization of `paginate` was moved to be after `super()`.

#### Changes Affecting ESM access

A `appName` parameter must be supplied to the constructor.

# Deprecations
The following sections detail parts of Backdraft that are deprecated and will be removed in future releases.

## Base and Listing plugins
All of these are deprecated:
`app.view()` - subclass `Backdraft.View`.
`app.view.listing()` - subclass `Backdraft.Listing.ListView`
`app.view.listing.item()` - subclass `Backdraft.Listing.ItemView`

## DataTable plugin namespacing methods
`app.view.dataTable()` - Use `Backdraft.DataTable.Local` or `Backdraft.DataTable.ServerSide`
`app.view.dataTable.row()` - Use `Backdraft.DataTable.Row`
`app.view.dataTable.columnType()` - Pass a `Backdraft.DataTable.Config` with appropriate column types to instance of `Backdraft.DataTable.Local` or `Backdraft.DataTable.ServerSide`

### CODE REVIEW

File: examples_server/examples/data_table/basic/index.html [hunk 1]
```diff
@@ -27,7 +27,6 @@ <h2>Basic Table</h2>
     <div id="example-area"></div>
 
     <!-- libraries -->
-    <script type="text/javascript" src="../../../vendor/json2.js"></script>
```
Comment (nburwell) [inline examples_server/examples/data_table/basic/index.html]: so this PR shares a lot of changes that I also saw in https://github.com/Invoca/backdraft/pull/101 - is the plan to get 101 into master and rebase this one?  I'd be in favor of that (and I'm good if 101 goes to master now)

File: src/cache.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import _ from "underscore";
+
+function getKey(key) {
+  if (key.cid) return key.cid;
```
Comment (nburwell) [inline src/cache.js]: for net new code I think we're agreeing we'd prefer not to do one line if / return statements right?  (I know this was a code move, but just wanted to check and wondering if we should tweak hound for this

File: spec/legacy/listing_plugin_spec.js [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+describe("Listing Plugin", function() {
+
+  let app;
+  let baseExports;
+  let collection;
+
+  beforeEach(() => {
+    Backdraft.app.destroyAll();
+    app = Backdraft.app("myapp", {
```
[the remark this answers is not in the document]
Comment (dtognazzini) [reply spec/legacy/listing_plugin_spec.js]: We should use `this` vars everywhere. For this PR, I'm staying focused the objective. These tests are preexisting.

File: spec/legacy/listing_plugin_spec.js [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+      model : app.Models.M
+    });
+    collection = new app.Collections.col();
+    baseExports = Backdraft.plugin("Base");
+  });
+
+  describe("exports", function() {
+    let Listing;
+
+    beforeEach(() => {
```
[the remark this answers is not in the document]
Comment (dtognazzini) [reply spec/legacy/listing_plugin_spec.js]: Fixed.

File: spec/router_spec.js [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      afterSwap : spy.afterSwap
+    });
+
+    router.swap(view);
+
+    expect(spy.beforeSwap).toHaveBeenCalledWith(router);
+    expect(spy.afterSwap).toHaveBeenCalledWith(router);
+  });
+
+  it("should throw and error if $el is not found", function() {
```
[the remark this answers is not in the document]
Comment (dtognazzini) [reply spec/router_spec.js]: Yep. I'm avoiding making updates that are specific to the objective. This is preexisting.

File: spec/router_spec.js [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+      new Router();
+    }).toThrowError("$el can't be found")
+  });
+
+  describe("Named Routes", function() {
+    it("should create them based on callback name", function() {
+      class MyRouter extends Router {
+        get routes() {
+          return {
+            "files" : "index"
```
[the remark this answers is not in the document]
Comment (dtognazzini) [reply spec/router_spec.js]: I'll fix any of these that I added as part of this PR.

File: src/data_table/column_types/bulk.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  columnType.definition(function(dataTable, config) {
+    return {
+      bSortable: config.sort,
+      bSearchable: false,
+      sTitle: "<input type='checkbox' />",
+      sClass : "bulk",
+      mData: function(source, type, val) {
+        return dataTable.collection.get(source);
+      },
+      mRender : function(data, type, full) {
```
[the remark this answers is not in the document]
Comment (dtognazzini) [reply src/data_table/column_types/bulk.js]: I'll fix any of these that I added as part of this PR.

File: examples_server/examples/data_table/basic/index.html (hunk 1 above)
Comment (dtognazzini) [reply examples_server/examples/data_table/basic/index.html]: #101 is now merged to master. This branch was daisy-changed off #101's branch.

File: src/cache.js (hunk 2 above)
Comment (dtognazzini) [reply src/cache.js]: > for net new code I think we're agreeing we'd prefer not to do one line if / return statements right?

I think so. Are agreements are captured in invoca/style-guide.

**Comment by dtognazzini** [comment]
@nburwell @garvank The changes here are no longer WIP! :-)  -> They are ready for a full review!

I've made an effort to document all the changes in the description. There is a branch on Web that proves these changes are working.

File: spec/listing/listing_spec.js [hunk 8]
```diff
@@ -1,97 +1,88 @@
...
-    baseExports = Backdraft.plugin("Base");
+    collection = new Collection();
   });
 
   describe("exports", function() {
-    var Listing;
-
-    beforeEach(function() {
-      Listing = Backdraft.plugin("Listing");
+    it("internal classes are available on Backdraft.Listing", function() {
```
Comment (garvank) [inline spec/listing/listing_spec.js]: Grammar doesn't make sense here. Maybe rename to `it("should make internal classes available on Backdraft.Listing", function() {});`

File: spec/listing/listing_spec.js [hunk 9]
```diff
@@ -1,97 +1,88 @@
...
+        }
 
-    it("requires that a collection be provided", function() {
-      expect(function() {
-        app.view.listing("abc", {});
-        new app.Views.abc();
-      }).toThrowError("A collection must be provided");
-    });
+        expect(new ChildItem()).toEqual(jasmine.any(Item));
+        expect(new ChildList({collection})).toEqual(jasmine.any(List));
```
Comment (garvank) [inline spec/listing/listing_spec.js]: Need spaces padding inside of `{}` like you've got on line 79. Not sure why ESL didn't pick this up.

File: spec/listing/listing_spec.js [hunk 10]
```diff
@@ -1,97 +1,88 @@
...
-      listing = new app.Views.Main({ collection: collection });
+      listing = new Main({ collection });
       listing.render();
     });
 
     describe("collection is reset", function() {
-
       beforeEach(function() {
-        collection.reset([ { name: "Bob"}, { name: "Joe" } ]);
+        collection.reset([ { name: "Bob" }, { name: "Joe" } ]);
```
Comment (garvank) [inline spec/listing/listing_spec.js]: Our styleguide says no spaces after `[` or before `]` in an array. Not sure why it isn't picked up by ESL.

File: spec/listing/listing_spec.js [hunk 11]
```diff
@@ -154,48 +141,37 @@ describe("Listing Plugin", function() {
...
         expect(listing.cache.size()).toEqual(1);
       });
-
     }
 
     describe("collection is removed from", function() {
-
       beforeEach(function() {
-        collection.reset([ { name: "Bob"}, { name: "Joe"} ]);
+        collection.reset([ { name: "Bob" }, { name: "Joe" } ]);
```
Comment (garvank) [inline spec/listing/listing_spec.js]: Same array comment here, also is a few other places in this diff.

File: spec/listing/listing_spec.js (hunk 9 above)
Comment (dtognazzini) [reply spec/listing/listing_spec.js]: I'll look into updating our EsLint config.

File: spec/listing/listing_spec.js (hunk 10 above)
Comment (dtognazzini) [reply spec/listing/listing_spec.js]: I'll look into updating our EsLint config.

File: spec/listing/listing_spec.js (hunk 11 above)
Comment (dtognazzini) [reply spec/listing/listing_spec.js]: I'll look into updating our EsLint config.

File: spec/listing/listing_spec.js (hunk 8 above)
Comment (dtognazzini) [reply spec/listing/listing_spec.js]: Yeah. I was thinking `it` might read as "its".

**Comment by dtognazzini** [comment]
@nburwell One last set of changes here for your review:
https://github.com/Invoca/backdraft/compare/c5c586de3babcd64aaa1091465608f5c9f5a36f4...6bc1477c6a1744a0d0b28c2fe57a75e17d32bdda

Summary:
- Refactor app, plugin, and namespace pieces to make them easier to test.
- Fix various test isolation issues that came with managing global state.
- Improved test coverage for app, plugin, namespace, local and server-side data tables.
- Expanded `Backdraft.app()` to accept an instance of `Backdraft.App`. When used in this fashion, the (deprecated) base plugin is not installed on the app.
- Add a `appInstance.installPlugin()` method that should be used in plugin initializers to ensure that plugins they depend on are installed on the target application.

And this concludes the 0.6.0 feature set. :-)

The only future changes to this branch will be to address code review comments or bugs found during testing.

File: webpack.config.js [hunk 12]
```diff
@@ -4,10 +4,12 @@ const webpack = require('webpack');
 // File filter for targeting specific spec files.
 const defaultSpecFileFilter = "/_spec\\.js$/";
 
+const isProduction = process.env.NODE_ENV === "production";
+
 module.exports = {
   entry: "./src/entry.js",
 
-  devtool: 'inline-source-map',
+  devtool: (isProduction ? 'source-map' : 'inline-source-map'),
```
Comment (garvank) [inline webpack.config.js]: Do we want source maps at all for production?

Comment (dtognazzini) [reply webpack.config.js]: `production` here is not the production Rails environment. Production here is how the deliverable for this repo will be generated when publishing the node module. In this case, it says, bundle an external sourcemap file with the JS file in`dist/`. Consumers, namely Web, will choose whether to use the sourcemap. In the case of Web, the soucemap will only be used in test and development Rails environments.

Comment (garvank) [reply webpack.config.js]: Gotcha, thanks for the explanation!

**Pull request merged**